### ISSUE TITLE
Dynamically change proxy in AsyncClient

### ISSUE THREAD

**Issue by rk2592000936** [encode/httpx#1376]
### Checklist

- [x] There are no similar issues or pull requests for this yet.
- [ ] I discussed this idea on the [community chat](https://gitter.im/encode/community) and feedback is positive.

### Is your feature related to a problem? Please describe.

All my requests should be send through different proxies which I have parsed from the internet. But AsyncClient handler doesn't support changing proxies dynamically. If I don't have valid proxy, should I need to recreate a httclient?

## Describe the solution you would like.

Add a api that can make AsyncClient's proxies change dynamically

**Comment by StephenBrown2** [comment]
How dynamic do you want to be? You can configure proxies based on domains already using a dict mapping them: https://www.python-httpx.org/advanced/#domain-routing and a more complex setup: https://www.python-httpx.org/advanced/#complex-configuration-example

**Comment by VeNoMouS** [comment]
@StephenBrown2 i think the issue is (as i came hutning as well) with requests, one can simply set the proxy dynamically after the session instance... via `session.proxies = {'https': '....'}` , is there a way one can do this with out having to pass it on the instance creation?

**Comment by VeNoMouS** [comment]
I mean... we could just access the "private" attribute and replicate what your already doing in the class...

```python
        self._proxies: typing.Dict[
            URLPattern, typing.Optional[httpcore.SyncHTTPTransport]
        ] = {
            URLPattern(key): None
            if proxy is None
            else self._init_proxy_transport(
                proxy,
                verify=verify,
                cert=cert,
                http2=http2,
                limits=limits,
                trust_env=trust_env,
            )
            for key, proxy in proxy_map.items()
        }
        self._proxies = dict(sorted(self._proxies.items()))
```

Seems messy to do it this way..

**Comment by tomchristie** [comment]
I'm going to close this issue off as I don't think there's enough clear context of motivation described here. Tho it's pretty similar to https://github.com/encode/httpx/issues/1374

If there's an unbounded number of differing proxies then you might want to use one-shot requests, eg. use `httpx.get(..., proxies=...)` directly. Alternately you could keep a mapping of different client instances. Or create a custom transport to route between them.

There are good design reasons why we choose not to support adding new transport instances on the fly, eg...

```python
with httpx.client(...) as client:
    client.proxies = ...  # It's unclear what `__enter__`/`__exit__` behaviour to expect on the proxy transport instances.
```

**Comment by yifeikong** [comment]
The use case is very clear to me -- crawlers. When using requests, I could do this:

```py
import requests
s = requests.Session()
proxy = get_random_proxy()
s.get(url, proxies={"http": proxy, "https": proxy})
```

But with httpx, the proxies parameter is only available in `httpx.Client/AsyncClient`, I can only set it once when creating the session. What we are trying to do is to use different proxy for each url we access, thus lower the chance of being blocked by the the target sites.

If that was sounds reasonable to you, please reopen the issue. Thanks.

**Comment by tomchristie** [comment]
> What we are trying to do is to use a different proxy for each url we access, thus lower the chance of being blocked by the target sites.

I mean that sounds like a *potentially?* crappy thing to be doing, but yup. 😬

So, the deal here, is that every time you're switching to a different proxy, you're no longer getting the same connection pool. (Which would be the same with *any* HTTP client lib.)

With HTTPX we're not stopping you from being able to do what you want here, but we are enforcing that each time you're doing that, you're doing it with a clean session/client instance, which at least makes it more visible that you're *not* getting any connection pooling benefits.

If you want this you'll just need to use a style like...

```python
proxy = get_random_proxy()
with httpx.Client(proxies={"http": proxy, "https": proxy}):
    client.get(url...)
```

Each time, or else wrap it up in a little convince function that you can call, so...

```python
def send_request(url, ...):
    proxy = get_random_proxy()
    with httpx.Client(proxies={"http": proxy, "https": proxy}):
        client.get(url, ...)

send_request(url, ...)
```

**Comment by yifeikong** [comment]
AFAIK, Yes, I can use `httpx.get(url, proxies=get_random_proxy()` to switch proxies for each url but lose the benefits of pooling.

> every time you're switching to a different proxy, you're no longer getting the same connection pool.

I think I did not make things clear here. By random, I do not mean totally different, maybe just `random.choice` from a fixed limited list of proxies. Will that benefit from connection pooling?

To make the use case less crappy and more realistic, let's see we are a search engine, and we limit our http requests to be rooted from only limited and published IPs(say 1.1.1.0/24) so that the target websites know it's indeed us(by the source IP). So the problem become that how to rotate/load-balance http requests to those 256 proxy servers?

So, actually, I think the solution can be one of them:

1. Accept a list of proxies for `httpx.Client` and rotate them automatically. The current `httpx.Client` only accept one proxy for the same host.
2. The method mentioned earlier, add `proxies` for `httpx.Client.request` and pool the connection for the same proxy server. This way, users can leverage httpx to recycle broken or unused connections.

Or I have to make my own Clients pool:

```py
clients = []
for proxy in proxy_servers:
    client = httpx.Client(proxies=proxy)
    clients.append(client)

def send_request(url):
    client = random.choice(clients)
    return client.get(url)
```

**Comment by tomchristie** [comment]
> Or I have to make my own Clients pool

Yes, exactly this. That'll get you what you're looking for, and will give you some level of connection pooling.
